File: Python/LeetCode/LeetCode3660.py
# https://leetcode.com/problems/jump-game-ix

class Solution:
    def maxValue(self, nums: List[int]) -> List[int]:
        n = len(nums)
        suffixMin = [float("inf")] * (n + 1)
        for i in range(n - 1, -1, -1):
            suffixMin[i] = min(nums[i], suffixMin[i + 1])
        ans = [0] * n
        start = 0
        currentMax = 0
        for i in range(n):
            currentMax = max(currentMax, nums[i])
            if i == n - 1 or currentMax <= suffixMin[i + 1]:
                for j in range(start, i + 1):
                    ans[j] = currentMax
                start = i + 1
                currentMax = 0
        return ans

File: Swift/LeetCode/LeetCode3660.swift
// https://leetcode.com/problems/jump-game-ix

class Solution {
    func maxValue(_ nums: [Int]) -> [Int] {
        let n = nums.count
        var minTable = nums
        var maxTable = nums
        var answer = [Int](repeating: 0, count: n)
        for i in 1..<n {
            maxTable[i] = max(maxTable[i - 1], nums[i])
            minTable[n - 1 - i] = min(minTable[n - i], nums[n - 1 - i])
        }
        var componentStart = 0
        for i in 0..<n {
            if i == n - 1 || maxTable[i] <= minTable[i + 1] {
                for j in componentStart...i {
                    answer[j] = maxTable[i]
                }
                componentStart = i + 1
            }
        }
        return answer
    }
}
